architecture/adr-010-messaging-layers.md

```diff
@@ -0,0 +1,61 @@
+<!-- Morgan Stanley makes this available to you under the Apache License, Version 2.0 (the "License"). You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0. See the NOTICE file distributed with this work for additional information regarding copyright ownership. Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License. -->
+
+# Architecture Decision Record: Message Router Abstraction Layers
+
+## Context
+
+The Message Router is one of ComposeUI's core components. 
+The functionality of the Router should be customizable and pluggable, therefore we separate
+the distinct layers and their APIs, while we provide a default implementation for all of them.
+This document describes the abstraction layers of the Message Router and how they interact with each other.
+
+## Decision
+
+The Message Router should be divided into three layers:
+* Messaging
+* Serialization
+* Channel
+
+### Messaging API
+
+The Messaging API is a high level, object based communication API with the following features.
+
+* Provides a publish-subscribe communication API
+    * `Task PublishAsync<TMessage>(string topic, TMessage message)`
+    * `Task<IObservable<TMessage>> SubscribeAsync<TMessage>(string topic)`
+    * `Task UnsubscribeAsync(string topic)`
```

Comment:
This feels unintuitive to me. To subscribe, I'd have to do this:
```cs
var subscription = (await messageRouter.SubscribeAsync(topic)).Subscribe(new Observable(msg => HandleMessage(msg)));
```
`IObservable<T>.Subscribe` returns an `IDisposable` for unsubscribing, so what's the `Unsubscibe(topic)` method doing here?

Reply:
I've changed it as I suggested, plus I wrapped the `TMessage` in a `RouterMessage` to include metadata.

Reply:
We can still pass an `IObserver` to subscribe, that can receive error and completion events.
```cs
Task<IDisposable> SubscribeAsync<TMessage>(IObserver observer);
```
OR we can still use observables and let users leverage reactive extensions like they used to:
```cs
var subscription = messageRouter.Topic<TMessage>(topic).Subscribe(observer);
```
Note that there's now a `System.Reactive.Async` package which has `IAsyncObservable` & `IAsyncObserver`, with LINQ for `IAsyncEnumerable` and everything you can imagine. So it might make sense to just have an API facede that provides the sync and async API for subscriptions:
```cs
public sealed class Topic<TMessage> : IObservable<TMessage>, IAsyncObservable<TMessage>
{ ... }
```